aten/src/ATen/native/cuda/JitLoops.cuh

```diff
@@ -75,7 +75,7 @@ void jitted_gpu_kernel(
     const std::string& f,
     at::cuda::jit::BinaryFuncVariant scalar_pos =
         at::cuda::jit::BinaryFuncVariant::NoScalar,
-    at::opmath_type<f_inputs_type> scalar_val = 0,
+    f_inputs_type scalar_val = f_inputs_type{0},
```

Comment:
It doesn't make sense for this to be of type `at::opmath_type<f_inputs_type>`.

Lower down the code, we cast it back to `f_inputs_type` while passing to `jitted_gpu_kernel_impl`.
(NOTE third argument to template parameters, which is the type of `scalar_val` : ref impl of `jitted_gpu_kernel_impl` below)
https://github.com/pytorch/pytorch/blob/be354d8139f70c5fabfce882671ad1aed057b1ba/aten/src/ATen/native/cuda/JitLoops.cuh#L137-L147

where `jitted_gpu_kernel_impl` is 

https://github.com/pytorch/pytorch/blob/be354d8139f70c5fabfce882671ad1aed057b1ba/aten/src/ATen/native/cuda/CUDAJitLoops.cuh#L226-L228

Reply:
scalar_val should be in opmath_type, it's a mistake that it is cast to f_inputs_type below, and it should be fixed (that's the only place btw, as far as I can see true compute type is handled correctly everywhere else, and in the kernel signature scalar_val is opmath_t). This allows us to compute things like `half_tensor + 70000` that would otherwise overflow.

Reply:
Actually, digging a bit more, this has to be `f_inputs_type`.

`jitted_gpu_kernel_impl` names this as `compute_type` incorrectly.
https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/CUDAJitLoops.cuh#L226-L228

It passes this argument to either `launch_jitted_vectorized_kernel` or `launch_jitted_unrolled_kernel`.

https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/CUDAJitLoops.cuh#L295-L297

https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/CUDAJitLoops.cuh#L253-L254

Both of these templates expect and treat the parameter as `f_inputs_type` where `compute_type` is passed.

https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/CUDAJitLoops.cuh#L57-L71

They also take care of casting `scalar_val` to corresponding `opmath_t`.

Also for the code for kernel generation in `launch_jitted_*` expects that parameter to be `f_inputs_type` and figure the `compute_type` accordingly

https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/aten/src/ATen/native/cuda/CUDAJitLoops.cuh#L89-L91

**If we pass the `compute_type` then we view the data incorrectly while loading and compute incorrect results (verified from test failures)**

**With this in mind, the code**
* updates `scalar_val` to be of type `f_inputs_type` in `jitted_gpu_kernel`.
*  updates the parameter name of `jitted_gpu_kernel_impl` to `f_inputs_type`
* removes the `opmath_t` casting of `scalar_val` in `opmath_jitted_gpu_kernel_with_scalars` as that doesn't really make sense.

Let me know if that makes sense.
Thanks!

Reply:
As I said earlier, this is wrong. We cannot cast scalar_val to f_inputs_type because this will lead to overflow (and eager mode kernels depend on the values not overflowing). This needs to be fixed.

Reply:
Have updated to preserve `scalar_val` to be `opmath_t`. Thanks!

Reply:
Thanks!